Move DDQ Backend functional tests to their own repo

quantumew opened the pull request:
Made some small changes after moving. Fixed some bugs and added a script to run the tests in the desired order.

quantumew committed:
Initial commit of functional tests

quantumew committed:
Updating tests and making small amendments.

Squashing commits while testing with Travis.
Add script directory in run tests

Fix require statement

Add plugin to dependencies

Ugh gotta shim mysql backend plugin since it isn't published

topics not topic

Fix require statement to grab the plugin running the test.

I did not author this so I was a bit confused. I see what is going on
now.

Updating tests. Fixing small issues.

quantumew committed:
Add some more info to the README

fidian commented on README.md:
```diff
@@ -0,0 +1,35 @@
+============================
+
+Contains functional tests meant to be used by DDQ backend plugins. One example is [DDQ MySql Plugin](https://github.com/tests-always-included/ddq-backend-mysql).
+
```
Extra newline above headings, except the first one.

quantumew replied:
Fixed

fidian commented on README.md:
```diff
@@ -0,0 +1,35 @@
+        - bash script/mysql-commands.sh
+    script:
+        - bash node_modules/ddq-backend-tests/run-tests.sh
+
```
You can borrow this newline from here if you like.  That is recycling the bytes on the computer.  ♲

quantumew replied:
Fixed

fidian commented on README.md:
```diff
@@ -0,0 +1,35 @@
+        - bash node_modules/ddq-backend-tests/run-tests.sh
+
+
+The `script/mysql-commands.sh` script runs a setup file for MySql, the contents are below. Also see the manual test config for the test configuration [here](script/manual-testing-config.js).
```
Why would any MySQL-specific things live within this repository?

quantumew replied:
They don't. These are just examples using MySql to illustrate how to use it. I will make that more clear.

fidian commented on run-tests.sh:
```diff
@@ -0,0 +1,15 @@
+set -e
+thisDir="${0%/*}"
+
+# Running all tests rather then exiting if one failed.
```
The success/failure status code won't be returned for the first two tests.  How would a CI system catch problems?

quantumew replied:
That was an old comment. It will exit now if tests fails.

fidian commented on script/polling-heartbeat-test.js:
```diff
@@ -0,0 +1,74 @@
+
+assert = require("assert");
+config = require("./manual-testing-config");
+Plugin = require("../../../");
```
Why do you assume the plugin is located here?

quantumew replied:
I am presuming this is required by the plugin and the setup is the same as the MySql plugin. What should I do?

fidian replied:
Wouldn't the current working directory be the folder of the plugin?

    Plugin = require(process.cwd());

quantumew replied:
That is much better. Fixed. 3381ac60a5429af742107958d22f07bb6c38123e

fidian commented on script/wrapped-message-tests.js:
```diff
@@ -0,0 +1,159 @@
+tests = [
+    {
+        name: "heartbeat",
+        query: "INSERT INTO ?? (hash, isProcessing, owner, messageBase64, message, topic) VALUES('123', true, ?, '456', 'message', 'Test Topic');"
```
This is extremely MySQL specific.  I don't see how it would be possible for the MongoDB-based plugin to benefit from this in any way.

quantumew replied:
Yea I noticed that. I didn't write it. Guess I will rewrite it.

quantumew replied:
Actually I am going to defer to you on this. What should I do instead? Should this test expect a certain test data that is inserted by the plugin before the tests are run?

fidian replied:
This is a functional test, yes?  I would use the plugin and configure it to use a low heartbeat.  Then simulate what the clients would do.  Requeue is pretty easy - insert message, get message, requeue message, then listen again and get the same message.  Heartbeat failure may be a bit worse.  You could add a message, listen, then kill the heartbeat somehow (perhaps just run these things in a subprocess and kill the subprocess), then see if there are messages.  There shouldn't be.  Then wait and check again and the message should be available.

fidian commented on script/wrapped-message-tests.js:
```diff
@@ -0,0 +1,159 @@
+        instance.stopListening();
+        data[fn]((err, fnData) => {
+            if (err) {
+                done(err);
```
This doesn't return.  Also, the rest of the code isn't contained within an `else`.  This will always say the test was successful.

fidian replied:
This identical problem is also elsewhere.  Please get them all.

quantumew replied:
Fixed

fidian commented on script/wrapped-message-tests.js:
```diff
@@ -0,0 +1,159 @@
+        wrappedMessageTest(tests[testCounter].name, tests[testCounter].query, done);
+    } else {
+        assert.equal(errCount, 0, `Error Count: ${errCount}`);
+        process.exit(0);
```
How is it that this passes without an eslint directive?

I bet you don't have the stock `.eslintrc.yaml`, `.eslintignore`, and `.gitignore` files.

quantumew replied:
I followed the correct code standard documented here and added the standard linter files http://eslint.org/docs/rules/no-process-exit

quantumew commented on run-tests.sh:
```diff
@@ -0,0 +1,15 @@
+set -e
+thisDir="${0%/*}"
+
+# Running all tests rather then exiting if one failed.
```
That is an old comment it does indeed fail if any of the tests fail.

quantumew commented on script/wrapped-message-tests.js:
```diff
@@ -0,0 +1,159 @@
+        wrappedMessageTest(tests[testCounter].name, tests[testCounter].query, done);
+    } else {
+        assert.equal(errCount, 0, `Error Count: ${errCount}`);
+        process.exit(0);
```
Indeed just copied the tests over. gitignore is there but not the linter files. Added them. I will add a directive for it. The way it was written it would just keep running runNext over and over again and would hang so I added that line instead of rewriting the test.

quantumew replied:
I take that back. While I am rewriting other stuff I might as well fix this. Was mostly trying to get this over with since it isn't my code.

quantumew committed:
Adding standard project files. Setup travis. And fix code review

This is to ensure coding standards are followed. Also added travis
config, eslint call in test script, and fix code review comments.

quantumew commented:
@fidian all fixes should be here 9fb00846f5f9237ee92cae3a86f0d72334f2421c. Had two questions about the correct alternatives. Like I said this was basically copied from DDQ. I fixed some bugs but I didn't write 99% of this so I will defer to you on the proper fixes.

fidian commented:
Would you update your badges to match this?

```
[![npm version][npm-badge]][npm-link]
[![Build Status][travis-badge]][travis-link]
[![Dependencies][dependencies-badge]][dependencies-link]
[![Dev Dependencies][devdependencies-badge]][devdependencies-link]
```

and

```
[codecov-badge]: https://img.shields.io/codecov/c/github/tests-always-included/ddq-backend-tests/master.svg
[codecov-link]: https://codecov.io/github/tests-always-included/ddq-backend-tests?branch=master
[dependencies-badge]: https://img.shields.io/david/tests-always-included/ddq-backend-tests.svg
[dependencies-link]: https://david-dm.org/tests-always-included/ddq-backend-tests
[devdependencies-badge]: https://img.shields.io/david/dev/tests-always-included/ddq-backend-tests.svg
[devdependencies-link]: https://david-dm.org/tests-always-included/ddq-backend-tests#info=devDependencies
[npm-badge]: https://img.shields.io/npm/v/ddq-backend-tests.svg
[npm-link]: https://npmjs.org/package/ddq-backend-tests
[travis-badge]: https://img.shields.io/travis/tests-always-included/ddq-backend-tests/master.svg
[travis-link]: http://travis-ci.org/tests-always-included/ddq-backend-tests
```

quantumew committed:
Using cwd to require plugin and updating badges.

quantumew committed:
Squashed commit of the following:

commit f57a11bf257d4772d026239ea31f65dda730c276
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Fri Jan 20 13:28:21 2017 -0600

    Adding multiple listener test and make tests run serially.

commit 0cfbd1e81d0aa2a273ddadbc0679e15272926726
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Thu Jan 19 12:39:07 2017 -0600

    Adding a test to ensure heartbeat time elapses before getting message.

commit 1732749aadfc343765726a2b56534e2264e0da48
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Thu Jan 19 10:01:59 2017 -0600

    Change method of testing to be more specific to scenarios.

    This will be far more valuable.

commit 91ce687a1a434ea47f466f82880c6222e3e6a642
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Mon Jan 16 13:50:04 2017 -0600

    Refactor the functional tests for DDQ

quantumew commented:
@fidian squashed commits of test rewrites. I will get another reviewer for both pull requests f67fc3c40cbd82e063763f8af23e38e3e657dc25

kyle-long commented on .travis.yml:
```diff
@@ -0,0 +1,10 @@
+    - "6.0"
+    - "6"
+before_install:
+    - "npm install -g npm@latest"
```
https://tests-always-included.github.io/style-guide/general/ says "The last line in a file shall have a newline after it."

quantumew replied:
I say it shan't. Just kidding fixed. One bit commit hash will be at the bottom.

kyle-long commented on README.md:
```diff
@@ -0,0 +1,60 @@
+Usage
+-----
+
+Add this package as a devDependency and you can run the test script as part of your build process. You simply run the `run-test.sh` script. Here is an example with Travis and MySql.
```
Its supposed be cased like "MySQL".

I care very little about this

quantumew replied:
Fixed

kyle-long commented on README.md:
```diff
@@ -0,0 +1,60 @@
+        messageBase64 VARCHAR(256)
+    );
+
+Also see the manual test config for the test configuration [here](script/manual-testing-config.js).
```
I don't see this file.

kyle-long replied:
Discussed and this readme is just out of date.

quantumew replied:
Fixed

kyle-long commented on lib/index.js:
```diff
@@ -0,0 +1,107 @@
+     */
+    class BackendTester {
+        /**
+         * Constructs backend tester instance.
```
I don't think this comment is necessary.

quantumew replied:
I have to or the linter complains

kyle-long commented on lib/index.js:
```diff
@@ -0,0 +1,107 @@
+         * Constructs backend tester instance.
+         */
+        constructor() {
+            // I want to run all tests before throwing an error.
```
This is a good comment but the context it is in makes me a bit confused. Can we move this comment to the class description maybe? Just say "All tests will be run before erroring" or add additional details to this comment like "...which is why 'errored' was assigned to false" or whatever.

quantumew replied:
Fixed

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,361 @@
+    function multipleListenerTest(callback) {
+        instance.on("error", (err) => {
+            callback(err);
+            instance.removeAllListeners();
```
You might want this to go above the callback call. If it were me, I would expect my callback to be executed when everything is done.

quantumew replied:
Fixed the 3 instances of this

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,361 @@
+
+        console.log("Setting multiple listeners for test.");
+        accessed = false;
+        config.owner = "testerTwo";
```
FUNCTIONAL

This feels a bit dangerous. Every Plugin we create afterwards will be called testerTwo and this config is used globally.

I suggest making a copy
```
// Note: This is only a shallow copy. This should be changed if the config contains a sub object.
configTwo = Object.assign({}, config);
configTwo.owner = "testerTwo";
```

quantumew replied:
Indeed dumb, fixed

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,361 @@
+         */
+        function dealWithData(data) {
+            console.log(`Retrieved message on mutliple listener test ${data.message}`);
+            if (!accessed && data.message === "multiMsg") {
```
Consider wrapping this whole if else in another if and check `data.message` once there.

quantumew replied:
Lol totz. I must of added that later, how wonky

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,361 @@
+
+                // Making sure same message isn't access twice until requeued.
+                // Giving it a somewhat arbitrary amount of time.
+                setTimeout(() => {
```
This test makes an assumption that `instance` would always get the data first. the test wouldn't work the other way around. In fact I think it would hang forever if I am reading this correctly.

Perhaps we should change it to always take one action first, then the other second, regardless of instance consumes the message.

quantumew replied:
Yea I did that because I listen on one first so it would always get it first. I guess I shouldn't rely on that. Fixed

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,361 @@
+
+
+    /**
+     * Requeue test.
```
These comments are kind of pointless. May as well remove it or if you wanted to make it better you can go into more detail like "This is a test which will consume a message and then add it back into the queue".

quantumew replied:
Added detail

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @param {Function} callback
+     */
+    function requeue(callback) {
```
We discussed this sort of. It might just be me but renaming these may be beneficial. I keep getting them confused in my mind with the instances functions. This doesn't just run requeue but sets up a listener, waits to get the message and then requeues it.

quantumew replied:
Renamed

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,361 @@
+                        cleanupElapsed = true;
+                        console.log("Heartbeat lifetime time elapsed!");
+                    }, instance.config.heartbeatLifetimeSeconds * 1000);
+                    heartbeat((hbErr) => {
```
Already discussed but I got confused around this area. Maybe add a comment explaining that the heartbeat function sets up a listener and first "consumes" the message and THEN runs heartbeat on it so that when `remove` is called below and its listener is set up it will not get that message for the duration of the heartbeat.

I think I may have also been confused because this is exactly how you would NOT use it in a real situation. I imagine in a real situation you would set up something like.

```
setTimeout(() => {
    message.heartbeat().
}, 5000); // every five seconds.

// do work to use the message.
// Delete the message and clean up the timeout or something.
```

quantumew replied:
Added better comments

kyle-long commented:
Forgot to mention also, `run-tests.sh` still exists. I think it should be removed.

kyle-long commented:
One last thing. I think we should write a test that tests the whole deduplication part. Specifically I think it should test removal of those messages after they are consumed to make sure remove actually works in this scenario (I am skeptical).

quantumew committed:
Update readme, code review fixes, and add test.

We did not tests deduplication or the requeued functionality. I added a
test for that. I updated the README because the tests are no longer a
script but rather a module. Removed the run-tests script because it is
not used any longer.

Fix tests where there were missing returns and stopListenings.

This sets up many listeners so if we miss stopping one it will
drastically change the outcome of the test.

Adding deadlock count.

quantumew commented:
@kyle-long All code review fixes here 9df13d33a8ca3c2d18165819a59e836c3bb3c3c3

kyle-long commented on lib/index.js:
```diff
@@ -0,0 +1,107 @@
+         */
+        runAllTests(callback, timeout) {
+            if (!timeout) {
+                // 30 seconds should be more then enough time to run these
```
Could you update this comment?

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,405 @@
+            instance.on("data", (data) => {
+                instance.sendMessage("message", config.topics, (fail) => {
+                    if (fail && !fail.message === "Could not send message.") {
+                        callback(new Error("Was able to insert a duplicate message!"));
```
I don't get this message. It wouldn't have inserted a duplicate message in this scenario would it?

kyle-long commented on lib/tests.js:
```diff
@@ -0,0 +1,405 @@
+                                    callback(new Error("Duplicate message caused a requeue but message was not emitted for processing a second time."));
+                                }
+                            }, config.heartbeatLifetimeSeconds * 1000 + 1000);
+                            instance.on("data", (moreData) => {
```
I think this test is missing one more round of trying to consume the message. We don't know if the message is actually gone at this point right?

quantumew committed:
More code review. Make sure removal works and update comment.

quantumew committed:
Fixing flaw where callback was not getting called.

quantumew committed:
Make some more improvments.

quantumew committed:
Fixing commit.
